platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/migratable/FeatureProcessor.java

```diff
@@ -316,71 +420,141 @@ public void processUninstalledFeature(
-   * installed and then later started to get to the started state. Since we only want to deal with
-   * one state change at a time, we will purposely increase the number of attempts left for starting
-   * such that another processing round can be done at which point the state of the feature in
-   * memory will be seen as installed instead of uninstalled like it is right now such that it can
-   * then be finally started as it was on the original system. Otherwise, if it is not in the
-   * started state, it will simply be started.
+   * Processes the specified feature for starting.
+   *
+   * <p><i>Note:</i> If the feature was uninstalled, it will be installed and then later started to
+   * get to the started state. The change from installed to started will require another process
+   * attempt as we only want to deal with one state change at a time. The next processing round will
+   * see the state of the feature in memory as installed instead of uninstalled like it is right
+   * such that it can be finally started as it was on the original system. Otherwise, if it is not
+   * in the started state, it will simply be started.
    *
+   * @param jfeature the original feature information
    * @param feature the current feature from memory
-   * @param id the feature id
-   * @param state the feature state in memory
-   * @param region the region for the feature
    * @param tasks the task list where to record tasks to be executed
    */
-  public void processStartedFeature(
-      Feature feature, String id, FeatureState state, String region, TaskList tasks) {
+  public void processStartedFeatureToRecordTasks(
+      JsonFeature jfeature, Feature feature, TaskList tasks) {
+    final String id = feature.getId();
+    final FeatureState state = service.getState(id);
+
     if (state == FeatureState.Uninstalled) {
       // we need to first install it and on the next round, start it
-      // as such, let's make sure to increase the number of attempts left for the start operation
-      tasks.add(Operation.INSTALL, id, r -> installFeature(r, feature));
-      tasks.increaseAttemptsFor(Operation.START);
+      addCompoundInstallTaskFor(jfeature, tasks);
+      return;
     } else if (state != FeatureState.Started) {
-      tasks.add(Operation.START, id, r -> startFeature(r, feature, region));
+      // should use feature's region but we cannot figure out how to get it
+      tasks.add(Operation.START, id, r -> startFeature(r, feature, jfeature.getRegion()));
+    }
+    if (jfeature.isRequired() != service.isRequired(feature)) {
+      addCompoundUpdateTaskFor(jfeature, feature, tasks);
+    }
+  }
+
+  private void addCompoundInstallTaskFor(JsonFeature jfeature, TaskList tasks) {
+    // we shall verify if the installed feature is in the proper required state on a subsequent pass
+    // and if not, update its requirements appropriately
+    tasks.add(
```

Comment:
P.S. You used `jfeatures` in your snippet instead of `exportedFeatures` 😝 

I am not sure I understand your suggestion about having a factory method on TaskList ... by design the task list could not be qualified with a specific type of container where we accumulate because the container changes for each type of operation. For install it tracks JsonFeature objects and for uninstall it tracks Feature objects. In another PR, it is reused for bundles where it tracks a set of Bundle objects.  As such a factory method would most likely suffer the same problem as we have here. 

I like the idea of creating constants. Unfortunately, the second one is not a lambda but a closure as it depends on `jfeature` defined in the context of this function. The only way to achieve what you are proposing would be for me to define the second one as a static method where I can pass in `jfeature`. If you are fine with this I can surely do it. 
However, here is the downside to this ... we are increasing readability at the detriment of splitting apart the definition of two pieces of code that must be related together and to the third one; since one is creating the container for the other 2 methods. 

I definitely like the idea of the `createCompoundTaskIfAbsent()` with the notion of `adding` to it the subtasks so let me look into that. In such case, I think, the actual task (which you put in the `add`) should be part of the create call and the accumulation should part of a sequenced call to whatever is returned from the `createCompoundTaskIfAbsent()`.

Reply:
Yeah, didn't try the code myself before posting it so I'm not surprised there are some issues with it.

This was really just a suggestion to see if we could address some of the concerns we had about the original approach. Don't have a strong opinion on this one.

Reply:
Take a look at the modified version and let me know if you like that better.